### ISSUE TITLE
Instances and Plugins

### ISSUE THREAD

**Issue by tokejepsen** [pyblish/pyblish-base#40]
Im getting a bit confused as to how to handle instances in a validator. Currently I'm using the tagging of individual nodes, so the each instance is a node. The problem comes when the instance is a group or objectset.

How am I supposed to know what the instance is, without doing lots of checks in the plugin?

**Comment by mottosso** [comment]
An instance is supposed to be a single, self-contained, extractable unit - like a rig, model, animation, playblast or what not. That is, a collection of nodes along with their configuration.

So if you're tagging individual nodes, you probably can't rely on "family" as each "family" would represents an individual instance and you'd end up with one self-contained, extractable unit - per node; which is probably not what you want.

The way I've been approaching it, which seems to work, is by tagging a top-level node as being an instance and assuming all children to be part of that instance. In the case of objectSets, the objectSet is a top-level node, and all contained nodes are part of this instance. For transforms, I would assume the same is true.

So, in a validator or extractor:

``` python
class MyValidator(Validator):
    def process(self):
        for node in self.instance:  # Subclass of a Python set
            # This is where all the contained nodes are
            do_something(node)

        for key, value in self.instance.config.iteritems():  # Python dictionary
            # And this is where its configuration are at
            print "Maya attribute: {key} and value: {value}".format(key=key, value=value)
```

**Comment by tokejepsen** [comment]
ok, I could just have the transform selector put the single node into a list. Then we just have that as a rule for the selectors, that it should be iterable.

**Comment by mottosso** [comment]
I think what needs to happen is for this line.. 

https://github.com/abstractfactory/publish/blob/master/publish/plugins/select_transform.py#L17

..to not only add the node tagged with "publishable" and "family", but also all of its children:

``` python
for transform in cmds.ls("*." + publish.config.identifier,
                         objectsOnly=True,
                         type='transform'):

    instance = publish.domain.Instance(name=transform)
    for child in cmds.listRelatives(transform, allDescendents=True):
        instance.add(child)
```

Does that make sense? Then the `instance`, which is a set, will contain all nodes to be validated and extracted.

**Comment by tokejepsen** [comment]
That does make sense, although I'll need to see it in practice with tagging a single root node. Will have a go later tonight.

**Comment by mottosso** [comment]
It struck me that the current select() doesn't respect the context.
1. A context is created at [line 28 in main.py](https://github.com/abstractfactory/publish/blob/b6f5970549e72cbf0dcea533667ccf67c41929ff/publish/main.py#L28)
2. A new context is created at [line 37 in main.py](https://github.com/abstractfactory/publish/blob/b6f5970549e72cbf0dcea533667ccf67c41929ff/publish/main.py#L37)
3. Yet another context is created [within each plugin](https://github.com/abstractfactory/publish/blob/b6f5970549e72cbf0dcea533667ccf67c41929ff/publish/plugins/select_transform.py#L24).

What should really happen is for the initial call to `select()` to either be given a context, or create one. That context should then be passed along to plugins so that each plugin can append instances to it. At the moment, both `select_object_set.py` and `select_transform.py` overrides each other.

**Comment by tokejepsen** [comment]
Think you're right, was thinking of the context as a container for instances, but it also has the errors.

The Selectors should then be initialized with a context, similar to how the Validators are initialized with an instance. Maybe we can even initialize both Validators and Selectors with a context, and get that same base class?

**Comment by mottosso** [comment]
> Maybe we can even initialize both Validators and Selectors with a context, and get that same base class?

Ah, see that's where I'm running into issues. If a validator, say `validate_naming_convention.py` is given a Context, how will it know which instance to validate?

We might have to accept that they do very different things and possibly re-think how they correspond to the rest of the plugins. An alternative might be to pass more information along to each plugin, but I think that might be crowbaring a bit too much.

**Comment by tokejepsen** [comment]
> Ah, see that's where I'm running into issues. If a validator, say validate_naming_convention.py is given a Context, how will it know which instance to validate?

Shouldn't it process the instances that are tagged with the family the Validator operate on?

**Comment by mottosso** [comment]
I think I get what you're saying.. That would mean running the same Validator on all Instances, as opposed to passing each instance to its own Validator.

That could work!

**Comment by BigRoy** [comment]
Let me briefly elaborate on how I saw the concept of a _Context_ in relation to what I think it is in its current state in the project:

The _Context_ is the container for the data that travels through the _pipes_ (or _graph_ if you think of a node graph). A _Filter_ uses the data to do its processing and/or adds/alters data to the _Context_.

The thing I couldn't wrap my head around is that the **Instance** @mottosso was referring to the whole time was actually just a _piece of data_. In my mind it makes sense to look at it as if an _Instance_ is an _Attribute_ on a node. But when I look at the code every single node _selected from the scene_ seems to be considered an _instance_, though I would consider the a group of nodes to be a single _instance_ because it's a combined _piece of data_.

So I would expect something like:

```
nodes_in_scene = Instance()
timerange_in_scene = Instance()
```

And then a Validator should know (somehow) what it needs from the Context.

In its current stage I find it hard to figure out where the Validators are getting the timerange from the Context, and especially _how_ they are doing that. Or how we as a user of _Publish_ decide what _Instances_ from the _Context_ we do the validation on.

> Since I'm a Maya user I think of **Attributes** as _slots_ that hold data (inputs) or compute new data (outputs) that it then holds. This is a nice comparison because that is exactly how I think the **Instance** would operate within the **Context**.

The idea I had was to allow _Filters_ to specify _inputs_ and _outputs_ thus basically define the _attributes_ it has and which are considered the ones that it requires as input from the Context and which it will output. For example:

``` python
class MyValidator(Validator):
    def initFilter(self):
        self.addAttr("nodes", datatype=MayaNodes, input=True)
        self.addAttr("timeRange", datatype=Double2, input=True)
```

The above example defines two attributes that it requires as inputs. This also allows us to check whether the _set of filters_ that we combined can actually be processed before even running the operation. Because we will know whether we're getting the required information from the _Selectors_.
Also we _connect_ the attributes in the _node-graph_ or _pipes_ system so that we know what data ends up where.
